File: androidApp/src/main/java/dev/haroldjose/familysharedlist/android/presentationLayer/pages/navigator/viewmodels/NavigatorViewModel.kt
package dev.haroldjose.familysharedlist.android.presentationLayer.pages.navigator.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.navigation.NavHostController
import dev.haroldjose.familysharedlist.android.presentationLayer.pages.navigator.views.Router
import dev.haroldjose.familysharedlist.domainLayer.usecases.account.GetOrCreateAccountFromLocalUuidUseCase
import dev.haroldjose.familysharedlist.services.firebase.IFirebaseCrashlytics

class NavigatorViewModel(
    private val getOrCreateAccountFromLocalUuidUseCase: GetOrCreateAccountFromLocalUuidUseCase,
    private val crashlytics: IFirebaseCrashlytics,
): ViewModel(), INavigatorViewModel {
    override var viewState: NavigatorViewState by mutableStateOf(NavigatorViewState.Initial)

    override suspend fun checkIfNeedToCreateNewAccount(navController: NavHostController, initialRoute: Router) {
        viewState = NavigatorViewState.Loading
        try {
            getOrCreateAccountFromLocalUuidUseCase.execute()
            viewState = NavigatorViewState.Success
            navController.navigate(initialRoute.value)
        } catch (e: Throwable) {
            crashlytics.record(e)
            viewState = NavigatorViewState.Error(e.message ?: "Ocorreu um erro inesperado. Tente novamente.")
            return
        }
    }
}

File: iosApp/iosApp/presentationLayer/pages/navigator/viewModels/NavigatorViewModel.swift
import Foundation
import shared
import SwiftUI
import KMPNativeCoroutinesAsync

class NavigatorViewModel: NavigatorViewModelProtocol {
    @Published var viewState: NavigatorViewState = .loading

    private let getOrCreateAccountFromLocalUuidUseCase: GetOrCreateAccountFromLocalUuidUseCase
    private let crashlytics: IFirebaseCrashlytics

    init(
        getOrCreateAccountFromLocalUuidUseCase: GetOrCreateAccountFromLocalUuidUseCase,
        crashlytics: IFirebaseCrashlytics
    ) {
        self.getOrCreateAccountFromLocalUuidUseCase = getOrCreateAccountFromLocalUuidUseCase
        self.crashlytics = crashlytics
    }

    @MainActor
    func checkIfNeedToCreateNewAccount() async {
        viewState = .loading
        do {
            _ = try await asyncFunction(for: getOrCreateAccountFromLocalUuidUseCase.execute())
            viewState = .success
        } catch {
            showError(e: error)
            return
        }
    }

    func showError(e: Error) {
        crashlytics.record(error: e)
        viewState = .Error(message: e.localizedDescription, retryAction: {
            Task {
                await self.checkIfNeedToCreateNewAccount()
            }
        })
    }
}

